src/lib/vasi-sync/src/scchannel.rs

```diff
@@ -0,0 +1,254 @@
+            if state.contents_state == ChannelContentsState::Ready {
+                break;
+            }
+            if state.writer_died {
+                return Err(SelfContainedChannelError::WriterIsDead);
+            }
+            assert!(
+                state.contents_state == ChannelContentsState::Empty
+                    || state.contents_state == ChannelContentsState::Writing
+            );
+            let pre_sleeper_state = self.state.set_sleeper(sync::atomic::Ordering::Relaxed);
+            if pre_sleeper_state != state {
+                // Something changed; clear the sleeper bit and try again.
+                self.state.clear_sleeper(sync::atomic::Ordering::Relaxed);
+                continue;
+            }
+            let mut sleeper_state = pre_sleeper_state;
+            sleeper_state.has_sleeper = true;
+            let expected = sleeper_state.into();
+            match sync::futex_wait(&self.state.0, expected) {
+                Ok(_) | Err(nix::errno::Errno::EINTR) | Err(nix::errno::Errno::EAGAIN) => {
+                    // Something changed; clear the sleeper bit and try again.
+                    self.state.clear_sleeper(sync::atomic::Ordering::Relaxed);
+                    continue;
+                }
+                Err(e) => panic!("Unexpected futex error {:?}", e),
+            };
+        }
+        self.state.transition_contents_state(
+            ChannelContentsState::Ready,
+            ChannelContentsState::Reading,
+            sync::atomic::Ordering::Acquire,
+        );
+        let val = unsafe { self.message.get_mut().deref().assume_init_read() };
+        self.state.transition_contents_state(
+            ChannelContentsState::Reading,
+            ChannelContentsState::Empty,
+            sync::atomic::Ordering::Release,
+        );
+        Ok(val)
+    }
+
+    /// Marks the writer as dead, causing `receive` to not block, and to return
+    /// if it's already blocked. This is generally useful in conjunction with some type
+    /// of "watchdog" that detects that the writer has died, such as
+    /// `main::utility::ChildPidWatcher`.
+    ///
+    /// Panics if called more than once.
+    pub fn set_writer_dead(&self) {
+        let prev = self.state.set_writer_dead(sync::atomic::Ordering::Relaxed);
+        if prev.has_sleeper {
+            sync::futex_wake(&self.state.0).unwrap();
+        }
+    }
+}
+
+unsafe impl<T> Send for SelfContainedChannel<T> where T: Send {}
+unsafe impl<T> Sync for SelfContainedChannel<T> where T: Send {}
+
+// SAFETY: SelfContainedChannel is VirtualAddressSpaceIndependent as long as T is.
+unsafe impl<T> VirtualAddressSpaceIndependent for SelfContainedChannel<T> where
+    T: VirtualAddressSpaceIndependent
+{
+}
+
+impl<T> Drop for SelfContainedChannel<T> {
+    fn drop(&mut self) {
+        let state = ChannelState::from(self.state.0.load(sync::atomic::Ordering::Relaxed));
```

Comment:
I think this should be `Acquire` to make sure all previous changes to the message are visible here?